### DISCUSSION TITLE
Add an operator alike Promise.then with two arguments

### DISCUSSION THREAD

**Discussion by yastanotheruser** [ReactiveX/rxjs#7137] [Ideas / Feature request]
When migrating from Promises to RxJS observables it's recommended that `.then` chains be replaced with `.flatMap` operators or any others depending on the case. However, there's currently no operator that does what would be equivalent to `.then` with both `onFulfilled` and `onRejected` callbacks. I guess that would be equivalent to a `.map` operator that also takes two callbacks, one for `next` and other for `error`.

A pipe with both a sequence of `.map` operators and a `.catchError` operator would not behave the same, because errors thrown from the project function would be caught, whereas this does not happen when using a single `.then` with Promises.

**Comment by martinsik** [comment]
You can use `tap()` or handle the error in `subscribe()` handler. Evetually, you can "map" the error to a different value `catchError(error => of(whatever))` (even non-error value) or ignore it completely with `catchError(error => EMPTY)`.

**Reply by yastanotheruser** [reply]
`tap()` doesn't map the values or error, and as I said, chaining `map()` and `catchError()` is different because `catchError()` could catch an error thrown from the `map()` callback function. Here's an example:

```typescript
function getDataItems(): Promise<Data> {
  return firstValueFrom(this.http.get<Data>('/api/data'))
    .then((data) => {
      if (!responseIsValid(data)) {
        throw new InvalidResponseError('/api/data returned a bad response');
      }
      return data.items;
    }, (err) => {
      if (err instanceof HttpErrorResponse) {
        if (err.status === 404) {
          return [];
        }
      }
      throw err;
    });
}
```

For this trivial example I guess rethrowing in `catchError()` is the RxJS solution. Still it's currently not possible with operators to prevent catching the `InvalidResponseError`.

**Reply by martinsik** [reply]
`tap()` and `map()` are two different operators. You can use `map()` for error notifications. You want to have an operator that will catch only one specific type of error (like in your case `InvalidResponseError`)?

**Reply by yastanotheruser** [reply]
Not specifically, the example just illustrates how `.then()` with both fulfill and catch callbacks differs from `.then().catch()`, which maps to `.pipe(flatMap(), catchError())` for RxJS operators, but there's no equivalent of the former.
